### ISSUE TITLE
What's the meaning of frequency about optimizer

### ISSUE THREAD

**Issue by Maybewuss** [PyTorchLightning/pytorch-lightning#4456]
> The ‘frequency’ value is an int corresponding to the number of sequential batches optimized with the specific optimizer. It should be given to none or to all of the optimizers. There is a difference between passing multiple optimizers in a list, and passing multiple optimizers in dictionaries with a frequency of 1: In the former case, all optimizers will operate on the given batch in each optimization step. In the latter, only one optimizer will operate on the given batch at every step.

I don't know "only one optimizer will operate on the given batch at every step", if all optimizers's frequency is 1, which optimizer will be used?

**Comment by rohitgr7** [comment]
if you don't use frequency then all the optimizers with take a step in each `training_step`.
but if you define frequencies then it will follow the frequency intervals.
for eg:
if `opt_1_freq = 1 and opt_2_freq = 2`
the it will be like:
```
batch1 -> opt_1.step()
batch2 -> opt_2.step()
batch3 -> opt_2.step()
batch4 -> opt_1.step()
batch5 -> opt_2.step()
batch6 -> opt_2.step()
...
```
> if all optimizers's frequency is 1, which optimizer will be used?

in this case if will be like
```
batch1 -> opt_1.step()
batch2 -> opt_2.step()
batch3 -> opt_1.step()
batch4 -> opt_2.step()
...
```
it is sometimes used while training GANs.
see training procedure in `Improved Training of Wasserstein GANs`, Algorithm 1
paper: https://arxiv.org/abs/1704.00028

**Comment by Maybewuss** [comment]
> if you don't use frequency then all the optimizers with take a step in each `training_step`.
> but if you define frequencies then it will follow the frequency intervals.
> for eg:
> if `opt_1_freq = 1 and opt_2_freq = 2`
> the it will be like:
> 
> ```
> batch1 -> opt_1.step()
> batch2 -> opt_2.step()
> batch3 -> opt_2.step()
> batch4 -> opt_1.step()
> batch5 -> opt_2.step()
> batch6 -> opt_2.step()
> ...
> ```
> 
> > if all optimizers's frequency is 1, which optimizer will be used?
> 
> in this case if will be like
> 
> ```
> batch1 -> opt_1.step()
> batch2 -> opt_2.step()
> batch3 -> opt_1.step()
> batch4 -> opt_2.step()
> ...
> ```
> 
> it is sometimes used while training GANs.
> see training procedure in `Improved Training of Wasserstein GANs`, Algorithm 1
> paper: https://arxiv.org/abs/1704.00028

Thanks, if i set frequency in lr schedule, it works like optimizer?

**Comment by rohitgr7** [comment]
Somewhat similar but not exactly. In case of scheduler if you set a frequency then `scheduler.step()` will take place at frequency interval and all the schedulers with be independent of each other.
For eg
```
scheduler_1_freq = 2
scheduler_2_freq = 1
interval = 'step' for both
```
then it will be like
```
batch1 -> scheduler_2.step()
batch2 -> scheduler_1.step() and scheduler_2.step()
batch3 -> scheduler_2.step()
batch4 -> scheduler_1.step() and scheduler_2.step()
batch5 -> scheduler_2.step()
...
```

**Comment by Maybewuss** [comment]
> Somewhat similar but not exactly. In case of scheduler if you set a frequency then `scheduler.step()` will take place at frequency interval and all the schedulers with be independent of each other.
> For eg
> 
> ```
> scheduler_1_freq = 2
> scheduler_2_freq = 1
> interval = 'step' for both
> ```
> 
> then it will be like
> 
> ```
> batch1 -> scheduler_2.step()
> batch2 -> scheduler_1.step() and scheduler_2.step()
> batch3 -> scheduler_2.step()
> batch4 -> scheduler_1.step() and scheduler_2.step()
> batch5 -> scheduler_2.step()
> ...
> ```

Thanks!

**Comment by Maybewuss** [comment]
> Somewhat similar but not exactly. In case of scheduler if you set a frequency then `scheduler.step()` will take place at frequency interval and all the schedulers with be independent of each other.
> For eg
> 
> ```
> scheduler_1_freq = 2
> scheduler_2_freq = 1
> interval = 'step' for both
> ```
> 
> then it will be like
> 
> ```
> batch1 -> scheduler_2.step()
> batch2 -> scheduler_1.step() and scheduler_2.step()
> batch3 -> scheduler_2.step()
> batch4 -> scheduler_1.step() and scheduler_2.step()
> batch5 -> scheduler_2.step()
> ...
> ```

If i only use one lr scheduler, set frequency = 1 is same as frequency = any number?

**Comment by rohitgr7** [comment]
nope... in the case of schedulers all are independent. so for eg if you set `frequency=3` and `interval='step'`
then
```
batch1 -> no scheduler step
batch2 -> no scheduler step
batch3 -> scheduler.step()
batch4 -> no scheduler step
batch5 -> no scheduler step
batch6 -> scheduler.step()
...
```

**Comment by Maybewuss** [comment]
> nope... in the case of schedulers all are independent. so for eg if you set `frequency=3` and `interval='step'`
> then
> 
> ```
> batch1 -> no scheduler step
> batch2 -> no scheduler step
> batch3 -> scheduler.step()
> batch4 -> no scheduler step
> batch5 -> no scheduler step
> batch6 -> scheduler.step()
> ...
> ```

Wow~ i see, i think these explanation should be added to documents...

**Comment by rohitgr7** [comment]
already there for optimizers. https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#configure-optimizers

but I think can be improved for schedulers. Want to submit a PR?

**Comment by Maybewuss** [comment]
> already there for optimizers. https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#configure-optimizers
> 
> but I think can be improved for schedulers. Want to submit a PR?

How to improve？I think the document is not very clear.....

**Comment by rohitgr7** [comment]
Just checked again, I think it's already there for both optimizers and schedulers. Check both the NOTE sections.

**Comment by Maybewuss** [comment]
> Just checked again, I think it's already there for both optimizers and schedulers. Check both the NOTE sections.

Maybe my English is not very good..

**Comment by aegonwolf** [comment]
@rohitgr7 I do not find the documentation all that clear and I also noticed that while personally coding it by adding a loop like in the WGAN paper it works as expected, but using the frequency parameter, which according to the documentation should do the same (and according to you, something slightly different in my understanding), it seems it does not.

Is it possible to have the discriminator just do n sweeps over the same batch for every sweep the generator does?

**Comment by aaronmkts** [comment]
@aegonwolf did you manage to find a solution? Currently, want to implement multiple discriminator updates per gen update and am unsure if I should be using the frequency parameter in configuring optimizers or writing a for loop over the discriminator training.
